### ISSUE TITLE
Implement Conditional Update Queries for Data Fields Requiring Millisecond Accuracy

### ISSUE THREAD

**Issue by tn3rb** [eventespresso/event-espresso-core#306]
Because our data model system is currently following an Active Record pattern, we have been performing separate reads and writes when updating data. 
This can increase the possibility of race conditions having a detrimental effect on the accuracy of our data, and can even lead to our data being outright corrupted.
 
consider the following events happening in the following order:

- Request A reads value x from the db (x = 2)
- Request A adds 1 to value x and saves it back to the db (now x = 3)
- Request B reads value x from the db (x = 3)
- Request B adds 1 to value x and saves it back to the db (now x = 4)

Most queries only take a fraction of a second to run, but let's consider a scenario where a server is getting absolutely smashed and run those requests so close together that they are effectively happening at the same time:

- Request A reads value x from the db (x = 2)
- Request B reads value x from the db (x = 2)
- Request A adds 1 to value x and saves it back to the db (now x = 3)
- Request B adds 1 to value x and saves it back to the db (now x = 3)

now imagine that value x is the number of tickets being sold.

Although WPDB does not support it directly, we need to implement UPDATE queries that utilize the existing column data as part of the calculation.
ie: 

```sql
UPDATE t1 SET col1 = col1 + 1;
```

This would be an improvement over our existing queries because race conditions will not result in data overwrites because the new value is added to the existing column data.

- Request A reads value x from Col_X in the db (x = 2)
- Request A adds 1 to the value of Col_X (now x = 3)
- Request B reads value x from Col_X in the db (x = 3)
- Request B adds 1 to the value of Col_X (now x = 4)

This however still does not prevent race conditions that result in overselling because we can still have the following type of scenario where Col_X should not have been allowed to exceed a value of 3 (ie: 3 spaces available):

- Request A reads value x from Col_X in the db (x = 2)
- Request B reads value x from Col_X in the db (x = 2)
- Request A adds 1 to the value of Col_X (now x = 3)
- Request B adds 1 to the value of Col_X (now x = 4)

These kinds of queries can be further enhanced by implementing conditionals within them that only perform the update if a set of conditions is met.

For example:

The following query would only update the sold count for the specified ticket, if doing so does not exceed the total quantity available:

```sql
UPDATE `wp_esp_ticket`
SET `tkt_sold` = 
CASE
   WHEN (`tkt_sold` + 4) < `tkt_qty`
   THEN `tkt_sold` + 4
   ELSE `tkt_sold` 
END
WHERE `TKT_ID` = 10;
```

So if ticket 10 has a Qty of 25, and already has a sold count of 20, then `( 20 + 4 ) < 25` would equate to true and the update would proceed. 
The exact same query happening a millisecond later however would not be successful because `( 24 + 4 ) < 25` would equate to false.

Of course, ticket sales are not limited solely by the `tkt_qty` and need to take the ticket's corresponding datetime sold and reg limit fields into consideration, which in turn need to take all other related tickets into consideration, which is an extremely complex set of calculations. So because of this, we will likely need to pre-calculate the ticket availability and then pass that value into the query to do something like:

```sql
UPDATE `wp_esp_ticket`
SET `tkt_sold` = 
CASE
   WHEN (`tkt_sold` + 4) < {QTY}
   THEN `tkt_sold` + 4
   ELSE `tkt_sold` 
END
WHERE `TKT_ID` = 10;
```

This of course means we're back to having a race condition because calculating a value for `{QTY}` can not be done within the same query because that calculation is far too complex to be performed in SQL. As I have said before, the only absolutely guaranteed way to prevent a race condition and overselling is to form a single line in the physical world. Switching to conditional update queries will however prevent any kind of data overwrites from occurring and result in greater data accuracy. Ultimately it may be possible to reduce the race conditions even further using transactional queries but this does not necessarily solve the race conditions caused by having to perform any pre-calculations.

Next steps include:

 - deciding which fields would benefit from this type of query: ie: `tkt_sold`, `tkt_reserved`, `dtt_sold`, `dtt_reserved`
 - deciding whether logic for update queries should be added to `EEM_Base` or performed as raw SQL from the respective models (ie: `EEM_Ticket` and `EEM_Datetime`)
 - deciding how / where / when pre-calculations will be performed to minimize race conditions

**Comment by tn3rb** [comment]
@mnelson4 
Do you have any comments with regards to this?
If we were to implement conditional update queries like I have outlined, how would you like to proceed with that within the models?

Simplest / easiest way imo is to simply pass the full SQL to WPDB, but you may want to implement a way to build queries like that in a way that conforms with how the model system currently works. 

Thoughts?

**Comment by mnelson4** [comment]
Yeah I'd mostly build the SQL directly.
I'd probably use a few helper methods to get the table and column names though. Eg instead of the exact query you pasted above, I'd do

```
class EEM_Ticket...
{
...
public function countTicketsSold($ticket_id, $quantit_to_sell, $max){
$table_name = $this->table();
$tkt_sold_column = $this->field_settings_for('TKT_sold')->get_table_column();
global $wpdb;
$this->_do_wpdb_query(
    'query',
    array(
        $wpdb->prepare(
             "UPDATE `{$table_name}`
SET `{$tkt_sold_column}` = 
CASE
   WHEN (`{$tkt_sold_column}` + %d) < %d
   THEN `{$tkt_sold_column}` + %d
   ELSE `{$tkt_sold_column}` 
END
WHERE `TKT_ID` = %d;',
            $quantit_to_sell,
            $max,
            $quantit_to_sell,
            $ticket_id
       )
);
```

This way if the table or column name ever changes, this code doesn't need to be changed.
If we encounter this situation again, I'd be inclined to add a helper function in `EEM_Base` then, but for now, assuming you agree this is unlikely to come up again soon, I think just adding a function to `EEM_Ticket` or something is ok.

**Comment by tn3rb** [comment]
right, ya I was just showing the basic query and used hardcoded values where we wouldn't in actual code.
Thanks for fleshing out the rest of that for me though :)

We definitely want to use this method of updating for the following table fields:

- esp_ticket.TKT_sold
- esp_ticket.TKT_reserved
- esp_datetime.DTT_sold
- esp_datetime.DTT_reserved

but there may be some others that would benefit from a more accurate approach to updating, such as some of the meta table fields.

**Comment by garthkoyle** [comment]
This is an issue on ES: https://github.com/eventespresso/eventsmart.com-website/issues/284

**Comment by mnelson4** [comment]
Related helpscout discussion: https://secure.helpscout.net/conversation/722796917/594977/#thread-2009994962

**Comment by mnelson4** [comment]
>  the only absolutely guaranteed way to prevent a race condition and overselling is to form a single line in the physical world...

I had an idea for something similar. Maybe we could store a "lock" in the database, so that only one session at a time can linearize the registration process.

Eg, just before we check tickets sold and finalize a registration, we make sure only one request can enter that code block at a time. 
This is my pseudo-code:

```
/**
 * Tries to acquire lock. If somebody else has had the lock for too long, assume they've had an error and so we'll
 * take it anyway.
 * @return boolean
 */
function haveLock()
{
    global $wpdb;
    $time = current_time();
    $max_lock_time = 20;
    $threshold = $time - $max_lock_time;
    return $wpdb->query("update wp_options
        SET
            option_value = CASE
                WHEN option_value < '{$threshold}'
                THEN '{$time}'
                ELSE option_value
                END
        WHERE
        option_name='ee_reg_lock'");
}

/**
* Releases the lock (by setting the time to a really long time ago) to allow other requests into the critical code block.
*/
function releaseLock()
{
    global $wpdb;
    $wpdb->update(
        $wpdb->options,
        array(
            'option_value' => '2000-01-01 00:00:00'
        ),
        array(
            'option_name' => 'ee_reg_lock'
         ),
         array('%d'),
         array('%s')
    );
}

while(! haveLock())
{
    sleep(1);
}
// ok! We have the lock. No other request can be in this request concurrently
process_registration();
// All done. Let somebody else in now.
releaseLock();
```

This psuedocode obviously isn't complete- it should do something more graceful than wait forever (like wait for 10 seconds, then return saying "hey I'm waiting", and some javascript code can see that, inform the user, and try again). Also, `process_registration()` shows I don't know exactly where best to put this. I'm not sure where in the Ticket selector/SPCO code we could put this forced linearization. I'm thinking wherever we update the TKT_sold and TKT_reserved etc.

Lastly, we could try to avoid this unless it's necessary. Eg, we could only invoke this logic when there are only 50 tickets left (although it's a race condition to decide that, so maybe not.)

But does the gist of it seem reasonable? And around what code block(s) could it surround?

**Comment by tn3rb** [comment]
My comment regarding forming a "single line" was meant as more of a commentary on the difficulty with preventing race conditions in http requests and not as a desirable goal to aim for. 

The biggest problem with trying to enforce a "single line" approach for Event Espresso, assuming we keep the current default session expiration time of 60 minutes, is that it would reduce the throughput of our registration process down to something like 1 registration per hour. This is just not acceptable and would be a step backwards as far as I'm concerned. Some of the sites we have seen these potential overselling race conditions occurring on have been processing well over 1 registration **per minute**, so enforcing a "single line" would result in less than 1/60th of the throughput. Even reducing the session expiration to 15 minutes would only allow a throughput of 4 registrations per hour or 1/15th the throughput.

So regardless of whether any of the above actually becomes necessary, I still think it is imperative that we first improve some of our critical queries by implementing update queries that increment stored values directly instead of performing separate reads and writes. To correct for the complexities involved with calculating ticket sale limits due to the interaction of ticket quantities and datetime/venue capacities, we could if need be, perform a secondary read query after any updates to ensure that limits have not been exceeded, but if that has happened, then simply undo the previous update and inform the registrant that their requested ticket quantity is not available.

If that alone does not prove to be enough to prevent overselling race conditions on super busy sites, then I would prefer to implement a variation on "rate limiting" that would run as a new middleware class in our request stack (see: `/core/services/request/`) but track the number of registrations in progress. The benefit of this approach is that the number of concurrent registrations could be controlled via an admin option and allow customers with less than optimal hosting to reduce the strain on their servers to a level that is more manageable for them. **Then** if an admin wants or needs to, they could reduce the number of concurrent registration allowed to only 1.

Here's a relatively simple example of a rate limiting request stack middleware class https://github.com/rtheunissen/guzzle-rate-limiter but there are lots of other examples out there that we could use for "inspiration".

This of course is not too too far off from what you are suggesting, but just implemented in a different manner.

A rate limiting middleware service could also potentially work in conjunction with what is suggested in #344

**Comment by mnelson4** [comment]
Discussed it with Brent in slack, here's my understanding of the highlights:

- in the above, Brent thought I was suggesting only allowing one active transaction (ie a transaction with a non-expired session) at a time; but I was more suggesting only allowing one active REQUEST at a time- or even only allowing one process to be in a particular code block at a time.
- Variations of this "form a single line" approach exist: implement it as a middlware to linearize the entire request; implement it in the command bus to linearize only that one request; implement it elsewhere in the code to linearize just the affected code (eg in the ticket selector and SPCO where we update the ticket and datetime reserved/sold counts)
- The main troubles with all the "form a single line" approach are:
-- all code needs to go through the line... we need to find all code that's updating the datetime/ticket reserved/sold counts and use it there. So it's a game of whack-a-mole to get all of them. (But I guess that, even if we only do this in the ticket selector and SPCO, we would prevent 99% of concurrency issues)
-- it creates a bottleneck in the code which could affect performance, and could cause some requests to timeout because they're awaiting the lock
- Brent's originally suggested "update queries" is that we'd never set a datetime/ticket reserved/sold count, we'd always modify the value relative to its current value in the DB. This might not prevent some overselling (in cases where some code READ the value before that, and performed logic based on that) but at least it would keep the counts accurate. (Eg, it's possible we could oversell, but at least the count of how many we oversold by would be accurate).


Brent said:
> so here's my bottom line:
> - i still want us to implement update queries for some of our critical values
> - i think rate limiting / db locks could also be helpful to prevent race conditions (but I think it will prove to be a deceptively complex approach)

So we'll use this ticket for the update queries, and I can create another one for the "form a single line" approach. The two efforts aren't mutually exclusive.

So, regarding these update queries, Brent also suggested overriding methods like `EE_Ticket::set()` to prevent directly setting the reserved and sold counts, and instead adding methods like `EE_Ticket::reserve($number_to_reserve)` which would just update the ticket's reserved count in the DB (and update the model object with the new value IMO) and probably return whether it was successfully able to reserve the indicated number of tickets (and if not, client code could interpret that to mean there are no tickets left to reserve so abort ticket reservation logic).
I think that will be good, although we'll probably want to initially just deprecate the `EE_Ticket::set()` when used for `TKT_reserved` and `TKT_sold` etc.

**Comment by mnelson4** [comment]
so @tn3rb how would you like to proceed with this? Should you, me, or both of us implement this? I'd suggest that if you'd really rather focus on REM etc, I can take a stab at it and you can review; otherwise vice-versa.

**Comment by tn3rb** [comment]
ya I have more than enough on my plate with REM and this does directly involve the models, so it would be great if you could tackle it, thnx.

**Comment by mnelson4** [comment]
> next steps... deciding which fields would benefit from this type of query: ie: tkt_sold, tkt_reserved, dtt_sold, dtt_reserved

Yeah I think those would be the fields.

> deciding how / where / when pre-calculations will be performed to minimize race conditions

I'm thinking we'll mostly implement these changes inside `EE_Ticket::increase_reserved()`, `EE_Ticket::decrease_reserved()`, `EE_Ticket::increase_sold()`, `EE_Ticket::decrease_sold()`, and the corresponding methods from `EE_Datetime`.

One big issue for me is testing. It's hard to know if we've fixed anything unless there's a testcase backing it up. The work on this ticket won't prevent all overselling; it will just prevent the counts (eg ticket reserved or sold) from getting off from the intended values.

One way to test this would be to place WP actions inside the methods where we do the updating, and when we want a test case to simulate another request, have code listen during those actions and performs some actions during them. (I'm unaware of any other way to test reliably reproduce concurrency issues, as these depend on the underlying operating system's scheduling software.)

Diving into the code to see about the feasibility of that...

**Comment by mnelson4** [comment]
# Thinking out loud

Currently the call stack/order looks like this:
`FHEE__EE_Ticket_Selector___add_ticket_to_cart__ticket_qty` filter fired when a ticket amount selected, which calls
`EED_Ticket_Sales_Monitor::validate_ticket_sale()`, which calls
`EED_Ticket_Sales_Monitor::_validate_ticket_sale()`, which calls
`EE_Ticket::refresh_from_db()` to "make sure" it's up-to-date, then also calls `EE_Ticket::qty('saleable')` to get the count of tickets available (again directly from the DB), and exits if none are available. If any are available, it calls
`EED_Ticket_Sales_Monitor::_reserve_ticket()`, which calls
`EE_Ticket::increase_reserved()`.

`EE_Ticket::increase_reserved()` gets the reserved count from the model object,
saved an extra meta recording the count change, then calls
`EE_Ticket::_increase_reserved_for_datetimes()`, which loops through its datetimes and calls
`EE_Datetime::increase_reserved()` on each, 
which calculates the datetime's new reserved count (based on the model object's values), does `AHEE__EE_Datetime__increase_reserved`, then updates the model object's reserved count, 
and `EE_ticket::_increase_reserved_for_datetimes()` saves each datetime.
Then `EE_Ticket::increase_reserved()` calls
`EE_Ticket::set_reserved()`, which updates the reserved count on the model object, and then
`AHEE__EE_Ticket__increase_reserved` is fired.
Then `EED_Ticket_Sales_Monitor::_reserve_ticket()` saves the ticket.

So where could problems be occurring? 
During `EED_Ticket_Sales_Monitor::_validate_ticket_sales()`, right after it refreshes the model object from the DB and calculates the tickets available from the DB, those counts could be off as before `EE_Ticket::increase_reserved()` is called. Then when the new values are saved, they'll overwrite the writes that happened while `EE_Ticket::increase_reserved()` was doing all its logic.

## How to simulate the problem
Add a action in `EED_Ticket_Sales_Monitor::_reserve_ticket()` before it calls `EE_Ticket::increase_reserved()` which updates the ticket and datetime counts. Currently this should cause the counts to be off (the true count sold will be higher than what will be stored in the DB.)

## How the new work will fix this
The new work inside `EE_Ticket::increase_reserved()` won't redo any of the calculations (the only way to avoid race conditions in there would be "form a line", like Brent said). All we can do is: when we go to write the new `TKT_reserved`, `DTT_reserved`, etc amounts to the DB, instead of just writing the amount, bump the current amount in the DB. The DB's amount will be kept correct. There is also a built-in check for overselling (on the individual ticket or datetime) which could fail. If it does (meaning we've narrowly avoided overselling), stop the train.

Also, if we last-minute detect we would have oversold a ticket or datetime, we would ideally cascade undo the other bumps we just did.

Eg, consider a ticket with 3 datetimes. First we bump the 3 datetimes successfully, then when we go to bump the ticket, we detect it would have oversold. We need to decrease the 3 datetimes.
Or had the 3 datetime been unsuccessful, we would have needed to undo the 2 previous datetimes. This is basically a manual DB transaction. 

Also, after the successful reserved and sold count "bumps", we might want to update those values on the model objects. BUT we should be careful not to save those amounts, as it's possible they're stale and we'll be overwriting the DB's value.

So it would also be nice that, when saving, we only save those values upon initial insert or when we explicitly want to save them. (Eg, when updating the db row from the model object, we avoid setting those columns UNLESS some code explicitly called `set` for them.)


## Summarized plan

* [ ]  Add a action in `EED_Ticket_Sales_Monitor::_reserve_ticket()` and use to make a test to simulate overselling and erroneous ticket/datetime counts
* [ ] Use ticket bumps instead of read-and-writing to ticket/datetime reserved/sold counts.
* [ ] If a count bump fails, undo previous bumps and trigger sold-out response
* [ ] (Maybe after a count bump, avoid saving the ticket/datetime reserved/sold counts when saving the model object... theoretically this will be good but might be unnecessary as we're very unlikely to want to update anything else on tickets/datetimes during requests that increase those values)

**Comment by mnelson4** [comment]
* [ ] I should probably use the new "bump" methods for bumping sold too, although I'm not sure how to handle failures (should the ticket sale not get approved?)
* [ ] I should make a code snippet to also simulate last-minute sell-out during SPCO
* [ ] Analyze other code that changes reserved and sold counts, see if they should use the bump methods and how they should handle failures

**Comment by mnelson4** [comment]
This should be resolved by #874. There is still #818 for further work on avoiding race conditions, but hopefully this will make it unnecessary.
